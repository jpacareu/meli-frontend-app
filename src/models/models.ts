import { History } from 'history';

export interface IState {
	items: IItem[];
	searchBox: string;
	categories: string[];
	loadingSearch: boolean;
	detailProduct: {
		item: IDetailItem;
	} | null;
}

export interface IActions {
	search: (...rest: any) => {};
	getProductDetail: (...rest: any) => {};
	loading: (...rest: any) => {};
	received: (...rest: any) => {};
	setState: (...rest: any) => {};
	handleError: (...rest: any) => {};
}

export interface IStateHistory extends IState {
	history: History;
}

export interface IDetailItem {
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
	sold_quantity: number;
	description: string;
}

export interface IItem {
	id: string;
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
}

export type ICondition = 'new' | 'used';

export interface IPrice {
	currency: string;
	amount: number;
	decimals: number;
}

export interface IStore {
	Provider: any;
	connect: () => any;
	actions: IActions;
}
