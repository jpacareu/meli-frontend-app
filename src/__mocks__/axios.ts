const create = {
	get: jest.fn(),
	put: jest.fn(),
	interceptors: {
		request: {
			use: () => ({})
		},
		response: {
			use: () => ({})
		}
	}
};

export default {
	get: jest.fn(),
	create: () => create
};
