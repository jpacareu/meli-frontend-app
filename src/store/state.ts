import { IState } from '../models/models';

export const initialState: IState = {
	items: [],
	searchBox: '',
	categories: [],
	loadingSearch: false,
	detailProduct: null
};
