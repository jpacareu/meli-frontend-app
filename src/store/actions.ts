import axios from 'axios';
import { IState, IActions } from '../models/models';

const baseURL =
	process.env.NODE_ENV === 'production'
		? 'https://meli-backend-app.herokuapp.com'
		: 'http://localhost:4000';

const axiosInstance = axios.create({
	baseURL: baseURL + '/api/'
});

export const actionsCreators: IActions = {
	search: async (state: IState, actions) => {
		let resultState: any = {};

		if (state.searchBox === '') return resultState;

		await actions.loading();

		try {
			const result = await axiosInstance.get('/items', {
				params: {
					q: state.searchBox
				}
			});

			const { items, categories } = result.data;
			resultState = { items: items || [], categories: categories || [] };
		} catch (error) {
			await actions.handleError(error);
		} finally {
			await actions.received();
		}

		return resultState;
	},
	getProductDetail: async (state: IState, actions, id: string = '') => {
		let resultState: any = {};
		await actions.loading();
		try {
			const result = await axiosInstance.get(`/items/${id}`);
			const { categories, ...rest } = result.data;
			resultState = {
				detailProduct: rest,
				categories
			};
		} catch (error) {
			await actions.handleError(error);
		} finally {
			await actions.received();
			return resultState;
		}
	},
	loading: () => {
		return { loadingSearch: true };
	},
	received: () => {
		return { loadingSearch: false };
	},
	handleError: (state, actions, error) => {
		// Handle error mock
		return {};
	},
	setState: (state, actions, value = {}) => {
		return { ...value };
	}
};
