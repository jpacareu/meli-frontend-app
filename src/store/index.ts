import createStore from 'react-waterfall';
import { actionsCreators } from './actions';
import { initialState } from './state';

const config = {
	initialState: initialState,
	actionsCreators: actionsCreators
};
const store = createStore(config);

export const { Provider, connect, actions } = store;
