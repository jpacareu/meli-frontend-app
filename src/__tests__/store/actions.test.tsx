import renderer from 'react-test-renderer';
import * as React from 'react';
import { Provider, connect, actions as a } from '../../store';
import { IActions } from '../../models/models';
import { initialState } from '../../store/state';
import axios from 'axios';

jest.mock('axios');

const actions = a as IActions;
let MainComponent;
let App;
let tree;
let spy: any = jest.fn().mockImplementation(state => null);

describe('actions object', () => {
	beforeEach(() => {
		spy = jest.fn(state => null);
		MainComponent = connect(state => ({ ...state }))(spy);

		App = () => (
			<Provider>
				<MainComponent />
			</Provider>
		);
		tree = renderer.create(<App />);
	});

	afterEach(() => {
		spy.mockReset();
	});


	it('actions triggered and state updated', async () => {
		expect(tree.toJSON()).toMatchSnapshot();
	});
	
	it("action 'handleError' is triggered and state is NOT updated", async () => {
		await actions.handleError({});

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(expect.objectContaining(initialState));
	});

	it("action 'getProductDetail' is triggered and resolved", async () => {
		const prodId = '1';
		axios
			.create()
			// @ts-ignore
			.get.mockImplementation(() =>
				Promise.resolve({ data: { categories: ['Category 1', 'Category 2'] } })
			);

		await actions.getProductDetail(prodId);

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(
			expect.objectContaining({
				categories: ['Category 1', 'Category 2']
			})
		);
	});

	it("action 'getProductDetail' is triggered and 'detailProduct' provided", async () => {
		const prodId = '1';
		axios
			.create()
			// @ts-ignore
			.get.mockImplementation(() =>
				Promise.resolve({
					data: {
						description: 'Description'
					}
				})
			);

		await actions.getProductDetail(prodId);

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(
			expect.objectContaining({
				detailProduct: {
					description: 'Description'
				}
			})
		);
	});

	it("action 'search' is triggered and resolved", async () => {
		await actions.setState({ searchBox: 'Detergente' });

		axios
			.create()
			// @ts-ignore
			.get.mockImplementation(() =>
				Promise.resolve({
					data: {
						items: [
							{
								id: '1'
							},
							{
								id: '2'
							}
						],
						categories: ['Category 1', 'Category 2']
					}
				})
			);

		await actions.search();

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(
			expect.objectContaining({
				items: [
					{
						id: '1'
					},
					{
						id: '2'
					}
				],
				categories: ['Category 1', 'Category 2']
			})
		);
	});

	it("action 'search' is triggered and rejected", async () => {
		const newState = { initialState, searchBox: 'Detergente' };
		await actions.setState(newState);

		axios
			.create()
			// @ts-ignore
			.get.mockImplementation(() =>
				Promise.reject({
					data: {}
				})
			);

		await actions.search();

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(expect.objectContaining(newState));
	});

	it("should not trigger API call for action 'search' if empty string provided", async () => {
		const newState = { initialState, searchBox: '' };
		await actions.setState(newState);
		const spy = jest.fn();

		axios
			.create()
			// @ts-ignore
			.get.mockImplementation(spy);

		await actions.search();

		expect(spy).not.toHaveBeenCalled();
	});


	it("action 'loading' is triggered and state updated", async () => {
		await actions.loading();

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(
			expect.objectContaining({
				loadingSearch: true
			})
		);
	});

	it("action 'received' is triggered and state updated", async () => {
		await actions.received();

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(
			expect.objectContaining({
				loadingSearch: false
			})
		);
	});

	it("action 'setState' is triggered and state updated", async () => {
		await actions.setState({ searchBox: 'Text' });

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(
			expect.objectContaining({
				searchBox: 'Text'
			})
		);
	});

	it("action 'setState' is triggered and state NOT updated when no parameter provided", async () => {
		await actions.setState();

		const instance = tree.root.findByType(MainComponent).children[0];

		expect(instance.props).toEqual(
			expect.objectContaining({
				searchBox: 'Text'
			})
		);
	});
});
