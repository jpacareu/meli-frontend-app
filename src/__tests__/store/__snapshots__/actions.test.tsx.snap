// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions object actions triggered and state updated 1`] = `null`;
